crates/matrix-sdk/src/widget/matrix.rs

```diff
@@ -309,38 +318,137 @@ impl MatrixDriver {
+            .await
+            .map(|s| s.is_encrypted())
+            // Default consider encrypted
+            .unwrap_or(true);
+
+        if room_encrypted {
+            trace!("Sending to-device message in encrypted room <{}>", self.room.room_id());
+
+            //  Attempt to re-batch these by content to batch them up into a single request
+            let mut content_grouped: BTreeMap<
+                &str,
+                BTreeMap<OwnedUserId, Vec<DeviceIdOrAllDevices>>,
+            > = BTreeMap::new();
+
+            for (user_id, device_map) in messages.iter() {
+                for (device_id, content) in device_map.iter() {
+                    content_grouped
+                        .entry(content.json().get())
+                        .or_default()
+                        .entry(user_id.clone())
+                        .or_default()
+                        .push(device_id.to_owned());
+                }
+            }
 
-        response.map_err(Into::into)
+            // Encrypt and send
+            for (content, user_map) in content_grouped {
+                // let's get all the devices for this content
+                let mut device_set: BTreeSet<(OwnedUserId, OwnedDeviceId)> = BTreeSet::new();
+                for (user_id, device_id_or_alias) in user_map {
+                    for id_or_alias in device_id_or_alias {
+                        match id_or_alias {
+                            DeviceIdOrAllDevices::DeviceId(device_id) => {
+                                device_set.insert((user_id.clone(), device_id.clone()));
+                            }
+                            DeviceIdOrAllDevices::AllDevices => {
+                                // get all the devices and add them
+                                let user_devices =
+                                    client.encryption().get_user_devices(&user_id).await?;
+                                let devices = user_devices.devices();
+                                devices.for_each(|device| {
+                                    device_set.insert((
+                                        device.user_id().to_owned(),
+                                        device.device_id().to_owned(),
+                                    ));
+                                })
+                            }
+                        }
+                    }
+                }
+
+                let mut recipients = Vec::<Device>::new();
+                for (u, d) in device_set {
+                    let found_device = client.encryption().get_device(&u, &d).await?;
+                    let Some(device) = found_device else {
+                        warn!("Device {} not found for user {}", d, u);
+                        // device not found, skip it and log
+                        failures.entry(u).or_default().push(d);
+                        continue;
+                    };
+                    recipients.push(device);
+                }
```

Comment:
Once you have all the recipient devices collected, getting the failures would be a simple set operation.

```
let failures = recipient_list - recipient_devices
```